File: tools/test_runtime_fixture.py
"""统一 Python 发布门禁的测试夹具身份校验。"""
from pathlib import Path
import hashlib,json,os
ROOT=Path(__file__).resolve().parents[1]
def sha(path):
 h=hashlib.sha256()
 with path.open('rb') as f:
  for block in iter(lambda:f.read(1024*1024),b''):h.update(block)
 return h.hexdigest()
def runtime(kind='raw',selected=None,full=False,require_current_archive=False):
 if selected is None:
  pointer=ROOT/'app/build/test-runtimes/current.json'
  if not pointer.is_file():raise ValueError('Run python tools/prepare-test-runtime.py first')
  selected=json.loads(pointer.read_text(encoding='utf8'))[kind]
 directory=Path(selected).resolve();proof=json.loads((directory/'dsha-test-runtime.json').read_text(encoding='utf8'))
 expected=json.loads((ROOT/'tools/dsh-runtime/package.json').read_text(encoding='utf8'))['dependencies']['@deepseek-ai/dsh']
 actual=json.loads((directory/'node_modules/@deepseek-ai/dsh/package.json').read_text(encoding='utf8'))['version']
 if proof.get('kind')!=kind or proof.get('dshVersion')!=expected or actual!=expected or proof.get('lockSha256')!=sha(ROOT/'tools/dsh-runtime/package-lock.json'):raise ValueError('TEST_FIXTURE_IDENTITY_MISMATCH')
 for file,digest in {**proof.get('archiveRecipeInputs',{}),**proof.get('overlayInputs',{})}.items():
  if sha(ROOT/file)!=digest:raise ValueError('TEST_FIXTURE_PATCH_INPUT_CHANGED:'+file)
 if require_current_archive and (not proof.get('archiveCurrent') or proof.get('archiveSha256')!=sha(ROOT/'app/src/main/assets/dsh-runtime.bin')):raise ValueError('TEST_FIXTURE_ARCHIVE_STALE')
 if full:
  if not proof.get('moduleHashes'):raise ValueError('TEST_FIXTURE_FILE_PROOFS_MISSING')
  for file,digest in proof['moduleHashes'].items():
   if sha(directory/file)!=digest:raise ValueError('TEST_FIXTURE_BYTES_CHANGED:'+file)
 return directory

File: tools/test-runtime-fixture.mjs
// Current DSH test runtime selection. Old historical fixtures must opt in explicitly.
import fs from 'node:fs';
import path from 'node:path';
import { createHash } from 'node:crypto';
const root=path.resolve(import.meta.dirname,'..');
const sha=file=>createHash('sha256').update(fs.readFileSync(file)).digest('hex');
const json=file=>JSON.parse(fs.readFileSync(file,'utf8'));
export function testRuntime(kind='raw', environment=kind==='raw'?'DSHA_TEST_RUNTIME':'DSHA_TEST_RUNTIME') {
  const expected=json(path.join(root,'tools/dsh-runtime/package.json')).dependencies['@deepseek-ai/dsh'];
  let directory=process.env[environment];
  if(!directory){const pointer=path.join(root,'app/build/test-runtimes/current.json');if(!fs.existsSync(pointer))throw Error('Run python tools/prepare-test-runtime.py to prepare current test inputs');directory=json(pointer)[kind];}
  directory=path.resolve(directory);
  const actual=json(path.join(directory,'node_modules/@deepseek-ai/dsh/package.json')).version;
  if(actual!==expected)throw Error(`Wrong runtime generation: expected ${expected}, found ${actual} at ${directory}`);
  const proofFile=path.join(directory,'dsha-test-runtime.json');
  if(!fs.existsSync(proofFile)) {
    if(process.env.DSHA_ALLOW_UNVERIFIED_RUNTIME!=='1')throw Error('Missing reproducible fixture provenance: '+directory);
    return directory;
  }
  const proof=json(proofFile);
  if(proof.kind!==kind||proof.lockSha256!==sha(path.join(root,'tools/dsh-runtime/package-lock.json')))throw Error('Stale or wrong fixture kind: '+directory);
  for(const [file,digest] of Object.entries({...proof.archiveRecipeInputs,...proof.overlayInputs}))if(sha(path.join(root,file))!==digest)throw Error('Fixture patch inputs changed: '+file);
  if(!proof.moduleHashes)throw Error('Missing per-module fixture fingerprints: '+directory);
  for(const [file,digest] of Object.entries(proof.moduleHashes))if(sha(path.join(directory,file))!==digest)throw Error('Fixture module bytes changed: '+file);
  return directory;
}
export function requireNativeHost(runtime) {
  const proof=JSON.parse(fs.readFileSync(path.join(runtime,'dsha-test-runtime.json'),'utf8'));
  if(proof.platform.os!==process.platform||proof.platform.cpu!==process.arch)throw Error(`Native test platform mismatch: fixture ${proof.platform.os}/${proof.platform.cpu}, host ${process.platform}/${process.arch}`);
}
